x/launch/keeper/invariants_test.go

```diff
@@ -33,18 +33,29 @@ func TestInvalidChainInvariant(t *testing.T) {
 	ctx, tk, _ := testkeeper.NewTestSetup(t)
 	t.Run("valid case", func(t *testing.T) {
 		chain := sample.Chain(r, 0, 0)
+		campaign := sample.Campaign(r, 0)
 		chain.LaunchID = tk.LaunchKeeper.AppendChain(ctx, chain)
+		chain.CampaignID = tk.CampaignKeeper.AppendCampaign(ctx, campaign)
+		chain.HasCampaign = true
 		msg, broken := keeper.InvalidChainInvariant(*tk.LaunchKeeper)(ctx)
 		require.False(t, broken, msg)
 	})
 
-	t.Run("invalid case", func(t *testing.T) {
+	t.Run("invalid case - chain already launched", func(t *testing.T) {
 		chain := sample.Chain(r, 0, 0)
 		chain.LaunchTriggered = true
 		chain.LaunchID = tk.LaunchKeeper.AppendChain(ctx, chain)
 		msg, broken := keeper.InvalidChainInvariant(*tk.LaunchKeeper)(ctx)
 		require.True(t, broken, msg)
 	})
+
+	t.Run("invalid case - chain does not have a valid associated campaign", func(t *testing.T) {
+		chain := sample.Chain(r, 0, 0)
+		chain.HasCampaign = true
+		chain.CampaignID = 1000
+		msg, broken := keeper.InvalidChainInvariant(*tk.LaunchKeeper)(ctx)
```

Comment:
The chain is not set therefore the condition is not checked

The invariant is still broken because of the state above is reused

We should instatiate a new `tk` or defined another test function

Reply:
Yeah, fair! my bad

Reply:
Note though: we reuse same `tk` basically in every other invariant test. I think for some it makes sense, but it could not for some others.

Reply:
I think most of the cases we should have valid and invalid cases where it doesn't pose a problem but we can have a check

Reply:
I checked, looks like we are good. Perhaps if you could give a look yourself at `campaign`, it's the only one I haven't looked at.